### ISSUE TITLE
[1.7.11] Issue with parsing of regex in configuration (non-capturing group): reports invalid regex

### ISSUE THREAD

**Issue by reischapa** [docker-library/haproxy#99]
Hello,

I came across some behavior that I am finding quite odd, which I am somewhat convinced may be a bug on version `1.7.11`.

To reproduce, first take the following example `Dockerfile`:

```
FROM haproxy:1.7.11
COPY haproxy.cfg /usr/local/etc/haproxy/haproxy.cfg
```

Save it to a directory. In that same location, put the following `haproxy.cfg`:

```
frontend 80
  bind *:80
  acl has_test_header hdr_reg(cookie) testHeader=(?:1)
```

Then, issue `docker build . -t test01 --no-cache` inside that.

Trying to run the resulting image with `docker run test01` results in the following output:

```
<7>haproxy-systemd-wrapper: executing /usr/local/sbin/haproxy -p /run/haproxy.pid -db -f /usr/local/etc/haproxy/haproxy.cfg -Ds 
[ALERT] 199/121933 (7) : parsing [/usr/local/etc/haproxy/haproxy.cfg:3] : error detected while parsing ACL 'has_latest_cookie_forced' : regex 'testHeader=(?:1)' is invalid.
[ALERT] 199/121933 (7) : Error(s) found in configuration file : /usr/local/etc/haproxy/haproxy.cfg
[ALERT] 199/121933 (7) : Fatal errors found in configuration.
<5>haproxy-systemd-wrapper: exit, haproxy RC=1
```
Meanwhile, if I change the version of the `haproxy` image to `1.8` (or `1.9`, or any other version above `1.7.11`), I get a different output, that I think indicates that the acl was parsed successfully:

```
[WARNING] 199/122522 (1) : config : missing timeouts for frontend '80'.
   | While not properly invalid, you will certainly encounter various problems
   | with such a configuration. To fix this, please ensure that all following
   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
```

Can anyone help me understand why is version `1.7.11` behaving in this fashion? Thanks in advance.

**Comment by wglambert** [comment]
It's possible they changed regex parsing in later versions, but from searching through their extensive release notes I didn't find anything particularly striking, and their forums don't offer much for the error [`regex is invalid`](https://discourse.haproxy.org/search?q=regex%20is%20invalid)

And these repos aren't really for help with using the software in the images so you'd have better luck asking over at [Stack Overflow](https://stackoverflow.com/search?tab=newest&q=haproxy) or maybe a haproxy forum

**Comment by reischapa** [comment]
Hi. Thanks for your answer.

A `Dockerfile` + `haproxy.cfg` configuration docker image that used similar regex logic to the one shown was actively in use and working, with a regex written as exemplified, for the last 5 or 6 months.

The reason we stumbled across this was that we noticed that we were unable to to build the image locally (meaning, in our local development machines) and push it to our private Docker registry, and then select it in Rancher for use in one of our proxy services (in a testing environment), exactly because of this invalid regex error.

On the other hand, if we were to use one of our GitLab runners to build the image, then the image would work just fine.

We were finally able to reproduce the issue on the runner when we deleted the cached base `haproxy` image (which was 6 months old), thus forcing it to re-download an image matching `haproxy:1.7`, which ended up being `haproxy:1.7.11`, the same that was being downloaded when building locally.

In short, this is why I think this is a bug on `haproxy:1.7.11` : 

  - This setup was working in some other version of `haproxy:1.7.x` (the one that was downloaded when `haproxy:1.7.11` was not yet available, but matched `haproxy:1.7`. Unfortunately, I cannot tell you which one it was at this point, but I can get that information if you think it would help)

  - This setup works on at least one version _after_  `haproxy:1.7.11` (I have made a mistake: I have only tested this with `haproxy:1.9.8`, but I can test this with `haproxy:1.8` and others, if you'd like)

**Comment by yosifkit** [comment]
I would initially guess that your issue might be related to #97, but from the history of the 1.7/Dockerfile I think it could also be #91.

**Comment by wglambert** [comment]
Did you ever make any progress on this issue?

**Comment by reischapa** [comment]
@wglambert we just ended up editing the dockerfile to use one of the 1.9.x versions (not sure which one), which was enough to bypass the issue.

If you want to try to reproduce the problem I would suggest contacting @joaosamouco, he may be better suited to help you at this point.

**Comment by wglambert** [comment]
Alright going to close then since this seems to be an upstream issue that isn't present in later versions
